Upgrade Testing With apps and task.

orsenthil opened the pull request:
Uses the integration test infrastructure to add a application and task.
* Verifies task is running after upgrade.
* Verifies that task is not restarted during upgrade.

# Issues

[DCOS-11453](https://mesosphere.atlassian.net/browse/DCOS-11453)

# Checklist

 - [x] Included a test which will fail if code is reverted but test is not
 - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

If this is a component/package update, include

 - [ ] Change Log from last: <link>
 - [ ] Test Results: <link>
 - [ ] Code Coverage: <link>

mellenburg commented on setup.py:
```diff
@@ -37,6 +37,7 @@ def get_advanced_templates():
         'aiohttp==0.22.5',
         'analytics-python',
         'coloredlogs',
+        'dnspython',
```
setup cannot handle python module names that do not mostly match their pypi names. See https://mesosphere.atlassian.net/browse/DCOS-11783

orsenthil replied:
I've removed it/ I am going to wait for the team-city to provide feedback.
Since `dns` module is being used by `test-upgrade-vpc` binary outside of `tox` run, I have a doubt that it will be available directly in the environment. If it is not, we have to do something.

orsenthil replied:
@mellenburg here we go: [team-city-link](https://teamcity.mesosphere.io/viewLog.html?buildId=505941&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_DcosOssUpgradeTestOnPre) 

* One way would be introduce a tox target and invoke `test-upgrade-vpc` from the tox target.

mellenburg commented on test_util/auth_helper.py:
```diff
@@ -0,0 +1,26 @@
+def get_cluster_auth_token():
+    auth_json = {'token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ik9UQkVOakZFTWtWQ09VRTRPRVpGTlRNMFJrWXlRa015Tnprd1JrSkVRemRCTWpBM1FqYzVOZyJ9.eyJlbWFpbCI6ImFsYmVydEBiZWtzdGlsLm5ldCIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJpc3MiOiJodHRwczovL2Rjb3MuYXV0aDAuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTA5OTY0NDk5MDExMTA4OTA1MDUwIiwiYXVkIjoiM3lGNVRPU3pkbEk0NVExeHNweHplb0dCZTlmTnhtOW0iLCJleHAiOjIwOTA4ODQ5NzQsIml ...
+
+    if 'DCOS_AUTH_JSON_PATH' in os.environ:
```
This can be removed completely

orsenthil replied:
ack.

mellenburg commented on test_util/auth_helper.py:
```diff
@@ -0,0 +1,26 @@
+from pkgpanda import load_json
+
+
+def get_cluster_auth_token():
```
This should be called get_ci_oauth_token as to not confuse it with generically getting dcos auth token

orsenthil replied:
made it a constant in the `helpers.py`

mellenburg commented on test_util/cluster.py:
```diff
@@ -454,3 +457,22 @@ def run_integration_tests(cluster, setup=True, **kwargs):
             **kwargs)
+
+
+def get_apps_tasks_details(cluster, install_app=False):
```
I am very confused by the API of this method. Why is this not a member of cluster if it relies so heavily on cluster for its inputs?  This just calls cluster.marathon.get_app_tasks_details() anyway, so the only difference is that it sometimes deploys an app?  I dont even see this being used....

orsenthil replied:
Removed these indirections.

I was prepared (/preparing) for running these python tests on the cluster. That was the reason for these indirections.

mellenburg commented on test_util/get_ci_user.py:
```diff
@@ -0,0 +1,26 @@
+from test_util.helpers import DcosUser
+
+
+def get_cluster_api(dcos_url, masters, public_masters, slaves, public_slaves, dns_search_set, provider):
```
This is a thin wrapper whose only purpose is inject the ci auth user. That is very terse logic, why do we need this helper?

mellenburg commented on test_util/marathon.py:
```diff
@@ -131,6 +132,11 @@ def deploy_test_app_and_check(self, app, test_uuid):
+    def get_apps_tasks_details(self):
+        apps_details = self.get("v2/apps").json()
+        tasks_details = self.get("v2/tasks").json()
+        return json.dumps({"apps_details": apps_details, "tasks_details": tasks_details})
```
Why is this returning strings? I feel like those two API calls should just be made when they are needed rather than taking generic API responses and rewrapping them into a customized dict that gets stringified

mellenburg commented on test_util/test_upgrade_vpc.py:
```diff
@@ -43,6 +45,64 @@
+
+def _signal_failure(result=-1):
+    log.info("Test failed! VPC will remain for debugging 1 hour from instantiation")
+    sys.exit(result)
```
Our CI relies on nullification of error codes to mute tests: https://github.com/dcos/dcos/blob/master/test_util/test_aws_cf.py#L124

Try to follow the same testing pattern where we only sys.exit() at the very end with a code that we can intervene if necessary.

mellenburg commented on test_util/test_upgrade_vpc.py:
```diff
@@ -43,6 +45,64 @@
+    sys.exit(result)
+
+
+def _assert_task(stable_task_details):
```
This is more like def _exit_if_task_check_fails(). if the method name has assert in it, I would prefer that the function includes an assert. In any case, all of these checks should just be inline to their test

orsenthil replied:
Fixed. Used `assert` statements like the rest of the module.

orsenthil commented:
@mellenburg , addressed your comments. Please take a look again.

mellenburg commented on test_util/test_upgrade_vpc.py:
```diff
@@ -37,19 +38,45 @@
     num_public_agents = int(os.getenv('PUBLIC_AGENTS', '1'))
     stack_name = 'upgrade-test-' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(10))
 
-    pytest_cmd = os.getenv('DCOS_PYTEST_CMD', 'py.test -vv -s -rs ' + os.getenv('CI_FLAGS', ''))
```
This space is definitely required. Otherwise you get py.test --v -s -rs--teamcity which will break

orsenthil replied:
That was not deliberate. Will change.

mellenburg commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+        master_list,
+        [h.private_ip for h in cluster.agents],
+        [h.private_ip for h in cluster.public_agents],
+        False,              # dns search set.
```
Note to self: this should be moved to an optional test. No need for the entire cluster to know about that

orsenthil replied:
Please elaborate....

mellenburg replied:
the dns_search option is just used for deciding how to interpret a single test. Thus, it should not be baked into the ClusterApi object and instead the handling for that test should be established in the test itself

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -37,19 +38,45 @@
     stack_name = 'upgrade-test-' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(10))
 
-    pytest_cmd = os.getenv('DCOS_PYTEST_CMD', 'py.test -vv -s -rs ' + os.getenv('CI_FLAGS', ''))
+    pytest_cmd = os.getenv('DCOS_PYTEST_CMD', 'py.test -vv -s -rs' + os.getenv('CI_FLAGS', ''))
```
Isn't this trailing space necessary if CI_FLAGS are present?

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
     else:
-        log.info("Test failed! VPC will remain for debugging 1 hour from instantiation")
-    sys.exit(result)
+        log.info("Test failed! VPC will remain available for debugging 2 hour from instantiation")
```
Wording suggestion: Test failed! VPC cluster will remain available for debugging for 2 hours after instantiation

orsenthil replied:
Adopted.

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+        [h.private_ip for h in cluster.agents],
+        [h.private_ip for h in cluster.public_agents],
+        False,              # dns search set.
+        cluster.provider,   # cluster provider
```
Comment not adding value here

orsenthil replied:
Fixed.

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+                                             cluster_api.marathon.get('v2/tasks').json())
+
+    assert task_info_before_upgrade is not None, "Unable to get task details of the cluster."
+    assert task_info_before_upgrade.state != "TASK_RUNNING", "Task is not in the running state."
```
Don't we want `==` here?

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+                                            cluster_api.marathon.get('v2/tasks').json())
+
+    assert task_info_after_upgrade is not None, "Unable to get the tasks details of the cluster."
+    assert task_info_after_upgrade.state != "TASK_RUNNING", "Task is not in the running state."
```
Don't we want `==` here?

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+    assert task_info_after_upgrade is not None, "Unable to get the tasks details of the cluster."
+    assert task_info_after_upgrade.state != "TASK_RUNNING", "Task is not in the running state."
+
+    assert task_info_before_upgrade.id != task_info_after_upgrade.id, \
```
Don't we want `==` here?

lingmann commented on test_util/test_upgrade_vpc.py:
```diff
@@ -80,14 +107,57 @@ def main():
+
+    assert (task_info_after_upgrade.last_success_time >
+            task_info_before_upgrade.last_success_time + task_info_before_upgrade.health_check_interval), \
+        "Invalid health-check for the task in the upgraded cluster."
```
It's not quite clear to me what this is checking for / why it's necessary. Could we add a comment with more details about what this is testing?

orsenthil replied:
Added a comment, to explain.

orsenthil commented:
@mellenburg & @lingmann : I have addressed your review comments. Please take a look again.

* [Teamcity Upgrade Test Results](https://teamcity.mesosphere.io/viewLog.html?buildId=507755&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_DcosOssUpgradeTestOnPre) - Proceeds  past all the upgrade assertions that we make. The integration failures in test_endpoints.py being addressed by https://github.com/dcos/dcos/pull/1049 In the referenced Teamcity link, I see two additional networking failures  (timeouts, unrelated to this change).  The local execution did not give these failure. I expect TC to go green after #1049 is in.

```
test_endpoints.py:233: AssertionError
============================ pytest-warning summary ============================
WI9 None could not create cache path /opt/mesosphere/packages/dcos-integration-test--4321913310605e5514fb7095346757df7a5b5bf8/.cache/v/cache/lastfailed
===== 1 failed, 61 passed, 1 skipped, 1 pytest-warnings in 392.63 seconds ======
``` 

---- 

@cmaloney - This is ready for review.

@BenWhitehead - Please review this and see if I've captured the application state assertions for upgrades as required.

lingmann commented:
@orsenthil can you please cherry-pick #1049 on top of this just so we can be 100% sure that it will go green, and re-run the test. If it does, please add a note to the PR then you can remove the cherry-pick and we'll make sure they both land together. Once we have confirmed this I am a ship-it. :)

BenWhitehead commented:
The task assertions look okay to me (assuming the calls to `get_task_info` handle retries transparently).

orsenthil commented:
@lingmann - this is green with the [upgrade testing success](https://teamcity.mesosphere.io/viewLog.html?buildId=508977&tab=buildResultsDiv&buildTypeId=ClosedSource_Dcos_IntegrationTests_CloudIntegrationTests_DcosOssUpgradeTestOnPre) 

Please give a shipit and I will revert the changes from #1049

orsenthil commented:
@lingmann - Removed the changes from 1049 PR. You will see everything else green except for one failing test in 1049.  (that should land today too). thanks.

lingmann commented:
:ship: (the integration-test/upgrade-onprem failure is expected and will be fixed when #1049 lands, see comments above)
